grains/ec2_info.py

```diff
@@ -28,24 +28,48 @@ def _call_aws(url):
         return response.read()
 
 
-def _get_ec2_hostinfo(path="", data={}):
+def _get_ec2_hostinfo(path=""):
     """
     Recursive function that walks the EC2 metadata available to each minion.
     :param path: URI fragment to append to /latest/meta-data/
-    :param data: Dictionary containing the results from walking the AWS meta-data
 
-    All EC2 variables are prefixed with "ec2_" so they are grouped as grains and to
-    avoid collisions with other grain names.
+    Returns a nested dictionary containing all the EC2 metadata. All keys
+    are converted from dash case to snake case.
     """
-    for line in _call_aws("/latest/meta-data/%s" % path).split("\n"):
+    resp = _call_aws("/latest/meta-data/%s" % path).strip()
+    d = {}
+    for line in resp.split("\n"):
         if line[-1] != "/":
             call_response = _call_aws("/latest/meta-data/%s" % (path + line))
             if call_response is not None:
-                data["ec2_" + path.replace("/", "_") + line] = call_response
+                line = _dash_to_snake_case(line)
+                try:
+                    data = json.loads(call_response)
+                    if isinstance(data, dict):
+                        data = _snake_caseify_dict(data)
+                    d[line] = data
+                except ValueError:
+                    d[line] = call_response
             else:
-                data["ec2_" + path.replace("/", "_")[:-1]] = line
+                return line
```

Comment:
Looking back on this I'm worried this isn't actually a base case. Will look into this.